Trophy manager: Load data in another thread

MSuih opened the pull request:
Currently opening trophy manager causes the whole emulator to freeze as the window can't be displayed untill trophy data has been loaded from hard drive. This pause can last several seconds on larger trophy collections, making it look like that the emulator has crashed.

This PR shows an empty trophy manager and a loading message while the trophies are being loaded in background, and fills the manager with trophies once everything has been loaded. This makes the trophy manager feel more reponsive and should improve the user experience.

I tried to keep the git diff fairly small and now it's making me nervous whether this implementation is too simple. I'm open for suggestions on that.

Megamouse commented:
I'll review it later

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,16 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+
+	// Fill trophy list
+	auto loadProgress = new QProgressDialog(
+			"Loading trophy data, please wait...", QString(), 0, 1, this,
```
Text that is visible to the user must be used like this `tr("your text")` in order to make the application futureproof for localizations.

Megamouse replied:
Text that is visible to the user must be used like this tr("your text") in order to make the application futureproof for localizations. Excluded are log messages.

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,16 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+	auto loadProgress = new QProgressDialog(
+			"Loading trophy data, please wait...", QString(), 0, 1, this,
+			Qt::Dialog | Qt::WindowTitleHint | Qt::CustomizeWindowHint);
+	loadProgress->setWindowTitle("Loading trophies");
```
same with `tr("your text")`

Megamouse replied:
same with tr("your text")

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,16 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+	// Fill trophy list
+	auto loadProgress = new QProgressDialog(
+			"Loading trophy data, please wait...", QString(), 0, 1, this,
+			Qt::Dialog | Qt::WindowTitleHint | Qt::CustomizeWindowHint);
```
why double tabs?

MSuih replied:
Some code conventions recommend double indentation in instances like these so that they're easily distinguishable from standard indentation. I'll change it to single tabs if it's the style here.

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -581,18 +585,26 @@ void trophy_manager_dialog::ShowContextMenu(const QPoint& loc)
+
+	RepaintUI();
+	progressDialog->close();
+	delete progressDialog;
```
this doesn't actually fully remove the dialog.
use `progressDialog->deleteLater();` instead

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,16 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+	RepaintUI();
+
+	// Fill trophy list
+	auto loadProgress = new QProgressDialog(
```
i'd like to see that as a separate function, so that it can be called by other user interactions in the future.
Also, closing the application during the loading phase will result in an error. So please abort this thread before the trophy manager is destroyed.

MSuih commented:
Did the requested changes, including an std::atomic variable for holding the thread's state so that two threads can change it without issues. I'm not sure if this was what you had in mind but it felt like the cleanest way of handling it.

isJuhn commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -581,18 +586,39 @@ void trophy_manager_dialog::ShowContextMenu(const QPoint& loc)
 }
 
-void trophy_manager_dialog::PopulateTrophyDB()
+void trophy_manager_dialog::StartTrophyLoadThread() {
```
newline before {

isJuhn commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,17 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+	StartTrophyLoadThread();
+}
+
+trophy_manager_dialog::~trophy_manager_dialog() {
```
newline before {

JohnHolmesII commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,18 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+
+trophy_manager_dialog::~trophy_manager_dialog()
+{
+	if (m_thread_state == RUNNING) {
```
{ on newline
```suggestion
	if (m_thread_state == RUNNING)
	{
```

MSuih commented:
Added an option for cancelling the load process, which stops the load process and closes the window. Might be useful in case someone accidentally opens the dialog, and I don't see any reason not to implement it. Also fixed some of the indentation while doing that.

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -321,7 +321,20 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
+
+trophy_manager_dialog::~trophy_manager_dialog()
+{
+	if (m_thread_state == RUNNING)
```
This should be:
```suggestion
	if (m_thread_state != CLOSED)
```

Megamouse commented:
This still results in a lot of crashes and misbehavior.
The fix i mentioned above is necessary to fix the cancellation of the progress dialog (if you stick with your implementation).
There are still Problems when closing the app while the thread is running.
Also focusing on the main window while the thread is running will result in an error.
To my own disappointment I can't provide a proper fix without looking deeper into this.
Good luck, I'm happy to help out as much as I can at the moment.

MSuih commented:
Yeah you're right, for some reason the application crashes if focus switches from trophy window even for one second. I'll take a look at this tomorrow and see if I can figure out what's causing it, but I fear I might've taken too big of a task here. I've done some multithreading work in the past but Qt seems more confusing in that regard than other frontend frameworks.

I've already spotted some other potential trouble spots, I'll return in a day or two to report how far my research got me.

MSuih commented:
Had to spend some time figuring out various quirks but I think I managed to get a working implementation together. All UI-related functionality (aside from thread state) is now handled with signals and slots, which should fix the crashes. I coudn't trigger any on my system at least.

MSuih commented:
Can someone with Visual Studio check why the compiling isn't working? I don't have any issues with clion + msvc, so either Appveyor is doing something stupid or the problem is in solution files.

Megamouse commented:
it's quite simple.
A: your vcxproj file changes are wrong
B: you have to declare the Q_OBJECT at the beginning of the file

if you wish i can create a patch that should be applied on a previously cleaned commit (consisting only of your 2 trophy manager files)

MSuih commented:
I think I got a hang of it with the latest commit, but if it doesn't compile and/or you can tell it's oblivously wrong then please do provide some solution. I can't test the changes at all locally. I'll call it a night for today, but will check it out tomorrow.

And yeah, sorry about the Q_OBJECT stuff. I should've checked on how they were used here. Doesn't really matter, I don't think the moc fails on that.

Megamouse commented:
I pushed an updated commit that contains a few style fixes and also the fixed vs project files:
https://github.com/Megamouse/rpcs3/commit/511e0c6f049da27cf19dfeb03c195c570eff7365

Yet... there still is an error when I close the app before the thread is done:
![image](https://user-images.githubusercontent.com/23019877/48160916-798f1180-e2d9-11e8-80cc-c63783ebd748.png)

MSuih commented:
I've been busy today, so I'll look further into this tomorrow. But two questions, @Megamouse:

1. Do you want credit for the changes (seperate commit with author intact) or do I combine them into my commit?
2. Would it be enough to just wrap the call to LoadTrophyFolderToDB inside a try-catch? The fatal program exit seems to be a variant of std::exception from what I can tell. That way there would be a exception catching ready in case someone overhauled the method to handle corrupted trophies more sophisticatedly.
https://github.com/RPCS3/rpcs3/blob/de924eb15ed12c0956aca61be818b5b39178ce21/rpcs3/rpcs3qt/trophy_manager_dialog.cpp#L636

And I apoligize for the amount of work you have to do on my pr.

Megamouse commented:
I don't know if there's a better solution. we can try it for now if it works.
I don't mind the credit, just do what takes least amount of effort in terms of merging/commits.
I also don't mind the review work. What matters is that the feature works :)

Megamouse commented:
Remind me to review this again in 10 hours XD

MSuih commented:
How's the review process going, Megamouse?

Megamouse commented:
It seems to be fine now.
I still had a random crash but I couldn't reproduce it.
